File: pascal/day11_part1_2018.pas
program ChronalCharge;

var
  grid: array[1..300, 1..300] of Integer;
  x, y, serial, maxPower, maxX, maxY, i, j, sum: Integer;

begin
  assign(input, 'input.txt');
  reset(input);
  readln(serial);
  close(input);

  for x := 1 to 300 do
  begin
    for y := 1 to 300 do
    begin
      grid[x, y] := ((((x + 10) * y + serial) * (x + 10)) div 100 mod 10) - 5;
    end;
  end;

  maxPower := -46; // minimum possible value for a 3x3 square
  for x := 1 to 298 do
  begin
    for y := 1 to 298 do
    begin
      sum := 0;
      for i := x to x + 2 do
      begin
        for j := y to y + 2 do
        begin
          sum := sum + grid[i, j];
        end;
      end;
      if sum > maxPower then
      begin
        maxPower := sum;
        maxX := x;
        maxY := y;
      end;
    end;
  end;

  writeln(maxX, ',', maxY);
end.

File: training_data/reflection-tuning/day11_part1_2018.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const gridSize = 300

func calculatePowerLevel(x, y, serialNumber int) int {
	rackID := x + 10
	powerLevel := rackID * y
	powerLevel += serialNumber
	powerLevel *= rackID
	powerLevel = (powerLevel / 100) % 10
	return powerLevel - 5
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	serialNumber, _ := strconv.Atoi(scanner.Text())

	// Calculate power levels and create prefix sum grid
	grid := make([][]int, gridSize+1)
	for i := range grid {
		grid[i] = make([]int, gridSize+1)
	}

	for y := 1; y <= gridSize; y++ {
		for x := 1; x <= gridSize; x++ {
			power := calculatePowerLevel(x, y, serialNumber)
			grid[y][x] = power + grid[y-1][x] + grid[y][x-1] - grid[y-1][x-1]
		}
	}

	// Find the 3x3 square with the largest total power
	maxPower := -1000000
	maxX, maxY := 0, 0

	for y := 1; y <= gridSize-2; y++ {
		for x := 1; x <= gridSize-2; x++ {
			power := grid[y+2][x+2] - grid[y-1][x+2] - grid[y+2][x-1] + grid[y-1][x-1]
			if power > maxPower {
				maxPower = power
				maxX, maxY = x, y
			}
		}
	}

	fmt.Printf("%d,%d\n", maxX, maxY)
}
